mutiny-core/src/auth.rs

```diff
@@ -0,0 +1,215 @@
+#![allow(dead_code)]
+use crate::{
+    error::MutinyError,
+    lnurlauth::{make_lnurl_auth_connection, AuthManager},
+    logging::MutinyLogger,
+    networking::websocket::{SimpleWebSocket, WebSocketImpl},
+    storage::MutinyStorage,
+};
+use jwt_compact::UntrustedToken;
+use lightning::util::logger::*;
+use lightning::{log_error, log_info};
+use lnurl::{lnurl::LnUrl, AsyncClient as LnUrlClient};
+use reqwest::Client;
+use reqwest::{Method, StatusCode, Url};
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use std::sync::{Arc, RwLock};
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+struct CustomClaims {
+    sub: String,
+}
+
+pub(crate) struct MutinyAuthClient<S: MutinyStorage> {
+    auth: AuthManager<S>,
+    lnurl_client: Arc<LnUrlClient>,
+    url: String,
+    http_client: Client,
+    jwt: RwLock<Option<String>>,
+    logger: Arc<MutinyLogger>,
+}
+
+impl<S: MutinyStorage> MutinyAuthClient<S> {
+    pub fn new(
+        auth: AuthManager<S>,
+        lnurl_client: Arc<LnUrlClient>,
+        logger: Arc<MutinyLogger>,
+        url: String,
+    ) -> Self {
+        let http_client = Client::new();
+        Self {
+            auth,
+            lnurl_client,
+            url,
+            http_client,
+            jwt: RwLock::new(None),
+            logger,
+        }
+    }
+
+    pub async fn authenticate(&self) -> Result<(), MutinyError> {
+        self.retrieve_new_jwt().await?;
+        Ok(())
+    }
+
+    pub fn is_authenticated(&self) -> (bool, Option<String>) {
+        if let Some(ref jwt) = *self.jwt.try_read().unwrap() {
+            return (true, Some(jwt.to_string())); // TODO parse and make sure still valid
+        }
+        (false, None)
+    }
```

Comment:
don't think we need the bool here, is implicit from the Option being defined or not